### DISCUSSION TITLE
batch or bulk create

### DISCUSSION THREAD

**Discussion by Gcode808** [OpenTalker/SadTalker#271] [General]
is there a way to bulk create or batch process, like reading an input folder and generating in an output folder? is this possible?

**Comment by einarpetersen** [comment]
This seems like a good idea for creators looking to use this for serious content creation, I already have a fantastic idea that I would like to try out and bulk creation would be a brilliant addition where you could set all the parameters for the images and soundfiles to use etc. or generating them

**Comment by vinthony** [comment]
Hi, it is possible. 

You can manually write some bash code to run each file individually.

**Reply by einarpetersen** [reply]
Yeah I think most of us are just lazy and expect things handed on a silver platter :) ... Now that I've been able to run it locally on CPU using inference.py as you suggested I am actually considering doing just that because rendering one 7 minute piece takes about 14 hrs and risks getting killed before completion so I have decided to be running audio split into 1 - 1.5 minute segments as it works, so a script sounds like a good idea, so I can just set the computer to run continuously without me interfering.

**Reply by einarpetersen** [reply]
This is the initial CLI command I got working python3 inference.py --cpu --driven_audio /home/directory/audiofile.mp3 --source_image /home/directory/image.png

Below is a bash script that should accomplish the task (according to GPT-4). It works by looping through each image in the directory, and for each image, it loops through all the corresponding audio files. It then runs the command you specified for each pair of image and audio files.

```

#!/bin/bash

image_directory=/home/directory/images
audio_directory=/home/directory/audiofiles

# for each image file
for image_file in $(ls ${image_directory}/*.png | sort)
do
    # Get base image name without the directory and extension
    image_base=$(basename ${image_file} .png)
    
    # for each audio file corresponding to the current image
    for audio_file in $(ls ${audio_directory}/${image_base}_*.mp3 | sort)
    do
        # Run the command with the current pair of image and audio files
        python3 inference.py --cpu --driven_audio ${audio_file} --source_image ${image_file}
        
        # Wait for the command to finish before going to the next
        wait
    done
done

```
Please replace /home/directory/images and /home/directory/audiofiles with the actual directories of your image and audio files.

You can also make the script executable by running chmod +x script.sh, and run it by typing ./script.sh.

Please note that this script assumes that the image and audio files are named in a specific way, with the audio files having names that start with the name of the corresponding image file. If your files have different naming conventions, you may need to modify the script accordingly.

**Comment by TiE23** [comment]
Old discussion but I wanted to share that I bashed together a cheap batch solution along with random improvements in a fork in a branch:

https://github.com/TiE23/SadTalker/tree/my-fixes

It's quick and dirty.
![Screenshot 2024-10-08 122147](https://github.com/user-attachments/assets/751fe708-aaaa-44a6-8bbd-01b6b5ab8ae0)

### How to use Batch Mode
* Click the Batch Mode toggle button
* Define the input image(s)
  * If you have just one image, just place its path in the top Image Path input (double quotes are stripped automatically)
    * Hot tip: Use Ctrl+Shift+C to copy the path of a file when viewing it in Explorer
  * If you have multiple images, you can use new-line delimited entries with a string "code word"
    * The "code word" is a string that the batch process checks for in the audio input file's path so it knows to match a particular image to a particular audio file
    * For example, if we have Red, Purple, and Crown Caesars, and each Caesar has their own audio file with the string "purple", "red", or "crown" in its name (actually, path), you'll be able to match them by writing your code word + image path input as:
```
red,"C:\Some\Dir\imageA.png"
purple,"C:\Some\Other\Dif\imageB.png"
crown,C:\Yet\Another\Dir\crown-caesar.png
```
* Add the audio file paths
  * Provide the full path, simply adding a new line to delimit them
```
"C:\Audio\red-caesar.mp3"
C:\Other\Dir\crowny-line-2.mp3
"C:\Audio\purple-caesar.mp3"
```
![Screenshot 2024-10-08 121924](https://github.com/user-attachments/assets/0f1b0460-7e0b-4b87-8c44-d4c1bb7b77b4)

* Press generate and pray


### Personal Preference Adjustments
* I default to Full + Still mode, as it was my use case
* The results directory now names the files slightly better (no `##` anymore)
* The results directory now names the directories better. No more insane UUID, now it has the date and the name of the audio file you inputted (ex: `2024-10-08T13.15.30_line-5_take-2`)

### Extra feature
* Added a slider for "Expression Scale". I literally never bothered to actually test it, so, if it's broken, sorry

---
This was enough for my one-off use case and I do not plan to improve it. It might break if you screw up any input whatsoever (non-existent file, for example), and when the batch process is done it throws an error. That's how half-assed it is.

Also, please note that the script copies the image file (and maybe the audio, I can't remember) to a "working directory" and deletes that working copy. Just in case you have some kind of special circumstance where copying is bad.

If you think it needs work, please steal my code and make it better. ❤️

**Reply by workdojos** [reply]
Thanks for this, does it still work?

**Reply by TiE23** [reply]
Nothing's changed, so, yep!

**Reply by oneplacepress** [reply]
> Nothing's changed, so, yep!

For some reason no matter what I try I cant get the Batch Mode toggle to appear in the UI.
